dmoj/judge.py

```diff
@@ -175,107 +94,123 @@ def begin_grading(
-                    # 2. If the short_circuit flag is true, see <https://github.com/DMOJ/judge/issues/341>.
-                    if (batched_case.result_flag & Result.WA) and (not case.points or short_circuit):
-                        is_short_circuiting = True
-                    yield batched_case
-                yield BatchEnd()
-                continue
-
-            # Stop grading if we're short circuiting
-            if is_short_circuiting:
-                result = Result(case)
-                result.result_flag = Result.SC
-                yield result
-                continue
-
-            result = grader.grade(case)
-
-            # If the submission was killed due to an user-initiated abort, any result is meaningless.
-            if self._terminate_grading:
-                raise TerminateGrading()
-
-            # If the WA bit of result_flag is set and we are set to short-circuit (e.g., in a batch),
-            # short circuit the rest of the cases.
-            # Do the same if the case is a pretest (i.e. has 0 points)
-            if (result.result_flag & Result.WA) > 0 and (short_circuit or not case.points):
-                is_short_circuiting = True
-
-            yield result
-
-    def log_internal_error(self, exc=None):
-        # If exc is exists, raise it so that sys.exc_info() is populated with its data
-        if exc:
-            try:
-                raise exc
-            except:  # noqa: E722, we want to catch everything
-                pass
-        exc = sys.exc_info()
-
-        message = ''.join(traceback.format_exception(*exc))
+            self.current_judge_process = SubmissionJudgeProcess(
+                problem_id, language, source, time_limit, memory_limit, short_circuit, meta
+            )
 
-        # Strip ANSI from the message, since this might be a checker's CompileError
-        # ...we don't want to see the raw ANSI codes from GCC/Clang on the site.
-        # We could use format_ansi and send HTML to the site, but the site doesn't presently support HTML
-        # internal error formatting.
-        self.packet_manager.internal_error_packet(strip_ansi(message))
+            ipc_handler_dispatch = {
+                IPC_COMPILE_ERROR: self._ipc_compile_error,
+                IPC_COMPILE_MESSAGE: self._ipc_compile_message,
+                IPC_GRADING_BEGIN: self._ipc_grading_begin,
+                IPC_GRADING_END: self._ipc_grading_end,
+                IPC_GRADING_ABORTED: self._ipc_grading_aborted,
+                IPC_BATCH_BEGIN: self._ipc_batch_begin,
+                IPC_BATCH_END: self._ipc_batch_end,
+                IPC_RESULT: self._ipc_result,
+                IPC_UNHANDLED_EXCEPTION: self._ipc_unhandled_exception,
+            }
+
+            for ipc_msg in self.current_judge_process.run():
+                type, data = ipc_msg
+
+                handler_func = ipc_handler_dispatch.get(type)
+                if handler_func is None:
```

Comment:
Maybe `except KeyError` would be more pythonic here.

Reply:
Done.